src/exec/execute_pipeline.c

```diff
@@ -22,53 +22,67 @@
 #define PIPE_READ 0
 #define PIPE_WRITE 1
 
-int	execute_left_node(t_ast *node, t_ctx *ctx,
-		t_pipeline_conf *conf, int *pipe_fd)
+static t_pipe_conf	*construct_pipe_conf(void)
 {
-	int	ret;
+	t_pipe_conf	*conf;
 
-	conf->fd_out = pipe_fd[PIPE_WRITE];
-	ret = execute_ast_node(node, ctx, conf);
-	close(pipe_fd[PIPE_WRITE]);
-	return (ret);
+	conf = ft_xmalloc(sizeof(t_pipe_conf));
+	conf->prev_read = -1;
+	conf->prev_write = -1;
+	conf->next_read = STDIN_FILENO;
+	conf->next_write = -1;
+	return (conf);
 }
 
-int	execute_right_node(t_ast *node, t_ctx *ctx,
-		t_pipeline_conf *conf, int *pipe_fd)
+static void	update_pipe_conf(t_pipe_conf *conf, int pipe_fd[2])
 {
-	int	ret;
+	conf->prev_read = conf->next_read;
+	conf->prev_write = conf->next_write;
+	conf->next_read = pipe_fd[PIPE_READ];
+	conf->next_write = pipe_fd[PIPE_WRITE];
+}
 
-	if (conf->fd_in != STDIN_FILENO)
-		close(conf->fd_in);
-	conf->fd_in = pipe_fd[PIPE_READ];
-	if (node->type != AST_PIPE)
-		conf->fd_out = STDOUT_FILENO;
-	ret = execute_ast_node(node, ctx, conf);
-	close(pipe_fd[PIPE_READ]);
-	return (ret);
+static void	update_last_pipe_conf(t_pipe_conf *conf)
+{
+	conf->prev_read = conf->next_read;
+	conf->prev_write = conf->next_write;
+	conf->next_read = -1;
+	conf->next_write = STDOUT_FILENO;
 }
 
-int	execute_pipeline(t_ast *node, t_ctx *ctx, t_pipeline_conf *conf)
+void	close_pipe_fd(t_pipe_conf *conf)
 {
-	int	pipe_fd[2];
+	if (conf->prev_read != STDIN_FILENO)
+		close(conf->prev_read);
+	if (conf->next_write != STDOUT_FILENO)
+		close(conf->next_write);
+}
 
-	if (pipe(pipe_fd) == -1)
-	{
-		print_error("pipe", strerror(errno));
-		return (EXIT_FAILURE);
-	}
-	if (execute_left_node(node->left, ctx, conf, pipe_fd) == EXIT_FAILURE)
-	{
-		close(pipe_fd[PIPE_READ]);
-		close(pipe_fd[PIPE_WRITE]);
-		return (EXIT_FAILURE);
-	}
-	if (execute_right_node(node->right, ctx, conf, pipe_fd)
-		== EXIT_FAILURE)
+// TODO: Is it necessary to add guard to check if conf is NULL?
+int	execute_pipeline(t_ast *node, t_ctx *ctx, t_pipe_conf *conf)
+{
+	t_ast		*cur;
+	int			ret;
+	int			pipe_fd[2];
+
+	cur = node;
+	conf = construct_pipe_conf();
+	while (cur && cur->type == AST_PIPE)
 	{
-		close(pipe_fd[PIPE_READ]);
-		close(pipe_fd[PIPE_WRITE]);
-		return (EXIT_FAILURE);
+		if (pipe(pipe_fd) == -1)
+		{
+			print_error("pipe", strerror(errno));
+			free(conf);
+			return (EXIT_FAILURE);
+		}
+		update_pipe_conf(conf, pipe_fd);
+		ret = execute_ast_node(cur->left, ctx, conf);
+		if (ret == EXIT_FAILURE)
```

Comment:
does this line needs free ?
```suggestion
		if (ret == EXIT_FAILURE)
		free(conf);
```

Reply:
Exactly 💯